routing/bicycle_directions.cpp

```diff
@@ -133,86 +141,21 @@ void BicycleDirectionsEngine::Generate(RoadGraphBase const & graph, vector<Junct
-    bool const inIsForward = routeEdges[i - 1].IsForward();
-    UniNodeId const uniNodeId(inFeatureId, inSegId, inIsForward);
-
-    for (auto const & edge : outgoingEdges)
-    {
-      auto const & outFeatureId = edge.GetFeatureId();
-      // Checking for if |edge| is a fake edge.
-      if (!outFeatureId.IsValid())
-        continue;
-
-      FeatureType ft;
-      if (!GetLoader(outFeatureId.m_mwmId).GetFeatureByIndex(outFeatureId.m_index, ft))
-        continue;
-
-      auto const highwayClass = ftypes::GetHighwayClass(ft);
-      ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Error, ());
-      ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Undefined, ());
-      adjacentEdges.m_outgoingTurns.candidates.emplace_back(0.0 /* angle */, uniNodeId,
-                                                            highwayClass);
-    }
-
-    LoadedPathSegment pathSegment(UniNodeId::Type::Mwm);
-    // @TODO(bykoianko) This place should be fixed. Putting |prevJunction| and |currJunction|
-    // for every route edge leads that all route points are duplicated. It's because
-    // prevJunction == path[i - 1] and currJunction == path[i].
-    if (inFeatureId.IsValid())
-      LoadPathGeometry(uniNodeId, {prevJunction, currJunction}, pathSegment);
-
-    if (m_numMwmIds)
-    {
-      ASSERT(inFeatureId.m_mwmId.IsAlive(), ());
-      NumMwmId const numMwmId =
-          m_numMwmIds->GetId(inFeatureId.m_mwmId.GetInfo()->GetLocalFile().GetCountryFile());
-      pathSegment.m_trafficSegs = {{numMwmId, inFeatureId.m_index, inSegId, inIsForward}};
-    }
-
-    auto const it = m_adjacentEdges.insert(make_pair(uniNodeId, move(adjacentEdges)));
-    ASSERT(it.second, ());
-    UNUSED_VALUE(it);
-    m_pathSegments.push_back(move(pathSegment));
-  }
+  if (cancellable.IsCancelled())
+    return;
 
   RoutingResult resultGraph(routeEdges, m_adjacentEdges, m_pathSegments);
   RouterDelegate delegate;
-  Route::TTimes turnAnnotationTimes;
+  Route::TTimes dummyTimes;
   Route::TStreets streetNames;
 
-  MakeTurnAnnotation(resultGraph, delegate, routeGeometry, turns, turnAnnotationTimes,
-                     streetNames, trafficSegs);
-
-  // @TODO(bykoianko) The invariant below it's an issue but now it's so and it should be checked.
-  // The problem is every edge is added as a pair of points to route geometry.
-  // So all the points except for beginning and ending are duplicated. It should
-  // be fixed in the future.
-
-  // Note 1. Accoding to current implementation all internal points are duplicated for
-  // all A* routes.
-  // Note 2. Number of |trafficSegs| should be equal to number of |routeEdges|.
-  ASSERT_EQUAL(routeGeometry.size(), 2 * (pathSize - 2) + 2, ());
-  ASSERT_EQUAL(trafficSegs.size(), pathSize - 1, ());
+  MakeTurnAnnotation(resultGraph, delegate, routeGeometry, turns, dummyTimes, streetNames,
+                     trafficSegs);
+  CHECK_EQUAL(routeGeometry.size(), pathSize, ());
+  if (!trafficSegs.empty())
```

Comment:
Done:
```
  // In case of bicycle routing |m_pathSegments| may have an empty
  // |LoadedPathSegment::m_trafficSegs| fields. In that case |trafficSegs| is empty
  // so size of |trafficSegs| is not equal to size of |routeEdges|.
```